packages/@lwc/integration-karma/test/custom-elements-registry/index.spec.js

```diff
@@ -0,0 +1,317 @@
+// Error message differs between browsers
+const tagAlreadyUsedErrorMessage =
+    /(has already been used with this registry|Cannot define multiple custom elements with the same tag name|has already been defined as a custom element)/;
+
+function getCode(src) {
+    return fetch(src)
+        .then((resp) => resp.text())
+        .then((engineCode) => engineCode.replace(/process\.env\.NODE_ENV/g, '"production"'));
+}
+
+function getEngineCode() {
+    const engineDomSrc = document.querySelector('script[src*="engine-dom"]').src;
+    const promises = [getCode(engineDomSrc)];
+    if (!process.env.NATIVE_SHADOW) {
+        const syntheticShadowSrc = document.querySelector('script[src*="synthetic-shadow"]').src;
+        promises.unshift(getCode(syntheticShadowSrc));
+    }
+    return Promise.all(promises);
+}
+
+function createLWC({
+    tagName = 'x-foo',
+    skipInject = false,
+    text = 'Hello LWC',
+    globalLWC = 'LWC',
+    customElement = false,
+} = {}) {
+    // basic "Hello World" compiled LWC component
+    function tmpl($api) {
```

Comment:
Improvement for a future PR: Would be great to generated this than manually author it. So it won't start failing if the compiler output changed.

Reply:
Agreed, this is a bit fragile. I could not think of another solution for now though.

We also have some other tests that use this pattern FWIW.